cEP-0021.md

```diff
@@ -0,0 +1,379 @@
+
+> Currently default format is to strip the directory (`strip_dirs()`) and sort
+> the stats of profiler result by cummulative time (`cumtime`) and display the
+> profiler result on *CLI*.
+>
+> ```python
+> pstats.strip_dirs().sort_stats('cumtime')
+> ```
+
+After formatting the profile output Bear writer will have an option to
+either display the formatted ouput on terminal or dump the formatted result
+in a specified file.
+
+```
+coala -b Bear Mybear --files <file> --profile-bears "<filename or file-path>"
+```
+
+Bear writer will also have a option to direclty dump the profile output
+to a file which can be further use to create a Sunburn Diagram for better
+estimation of profiler stats with the help of external library like `snakeviz`.
+
+```
+coala -b Bear Mybear --files <file> --profile-bears "True,dump"
+```
+
+# Implementation Detail
+
+When user invoke the profiler with either `CLI` or by `coafile`
+then `execute` method of Bear Base Class will call the `run_bear_from_section`
+method
+
+```python
+def run_bear_from_section(self, args, kwargs):
+    ............
+    profile_bears_results, args, kwargs = self.setup_profiler(
+        *args, **kwargs)
+    return profile_bears_results if profile_bears_results else self.run(
+        *args, **kwargs)
+```
+
+which futher call the `setup_profiler()` method, and `setup_profiler()` method
+first create a `Profile`  ( name `prof`) then enable the profiler on `run()`
+method of bears then check whether keyword arguments passed to the Bear
+contain the `profile_bears` key or not.
+
+```python
+if profile_bears is None:
+    kwargs.pop('profile_bears')
+    return False, args, kwargs
+```
+
+- If `yes` then then it split all the value passed to
+  `--profile-bears` through CLI <br>i.e.
+  `--profile-bears "True,strip_dirs(),sort_stats('cumtime')"`
+  or by `coafile` and return a list of all the comma seperated values.
+  i.e. `['True','strip_dirs','sort_stats('cumtime')']`
+
+- If `No` then it will return `False` and no profiling will be done
+  and coala will continue it's execution.
+
+After that `setup_profiler` method check whether Bear writer directly want
+to `dump` the profile output <br>i.e. `--profile-bears "True,dump"`
+
+```python
+if u'dump' in profile_bears and u'True' in profile_bears:
+    kwargs.pop('profile_bears')
+    filename = '{}_{}.prof'.format(
+        kwargs['section_name'], kwargs['bear_name'])
+    open(filename, 'w+')
+    kwargs.pop('section_name')
+    kwargs.pop('bear_name')
+    cProfile.runctx('self.run(*args, **kwargs)',
+                    globals(), locals(), filename)
+    # cli.main(['output.prof'])
+    return False, args, kwargs
+```
+
+- If `yes` then it will create a profiler file for every bear of
+  every section and name of a `profiler file` naming convention will be like
+  `{section_name}{bear_name}.prof` so that if user want to make a sunburn
+  diagram of any bear of any section they can easily do that with `sankeviz`
```

Comment:
you should not leave it for the user. you have to create the sunburst diagram yourself if that is a proposed feature

Reply:
so if you are leaving it for the user then better not mention `snakeviz` at all as its up to the user to do whatever he/she wants with the `.prof` file

Reply:
It's better to remove the `snakeviz` part. :+1: